Sources/SwiftLexicalLookup/SequentialScopeSyntax.swift

```diff
@@ -0,0 +1,93 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Scope that, in addition to names introduced by itself,
+/// also handles names introduced by
+/// `IntroducingToSequentialParentScopeSyntax` children scopes.
+protocol SequentialScopeSyntax: ScopeSyntax {
+  /// Returns names introduced by `codeBlockItems`
+  /// and included `IntroducingToSequentialParentScopeSyntax` children
+  /// scopes that match the lookup.
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult]
+}
+
+extension SequentialScopeSyntax {
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
```

Comment:
It's here particularly because of the file scope. Since it's results are of `.fromFileScope` kind, we need a way to define which `LookupResult` should we use. Do you think there could be a cleaner way to provide this?

Reply:
Why do we need to differentiate between `fromScope` and `fromFileScope` at all? Couldn’t we conform `SourceFileSyntax` to `ScopeSyntax` and then consider the cases the same?

Reply:
It was a design decision we agreed on with @DougGregor to further differentiate the weird behavior of file scope from the rest of the scopes. `SourceFileSyntax` already conforms to `ScopeSyntax`, but I think it's nice to have it's results separated.

Reply:
Part of the reason for having `fromFileScope` be separate is because clients need to do extra work to do lookups into other source files and in visible imports from that scope, so it's both "here's what I found in this file" and "you need to go look elsewhere, too". I'm not totally settled on what the right answer is for dealing with file scope, so I'd rather keep thinking about it (but not block this PR on a resolution).